scipy/optimize/tests/test_lbfgsb_setulb.py

```diff
@@ -114,3 +114,11 @@ def test_setulb_floatround():
 
         assert (x <= upper_bnd).all() and (x >= low_bnd).all(), (
             "_lbfgsb.setulb() stepped to a point outside of the bounds")
+
+
+def test_gh_issue18730():
+    def fun(x):
+        return np.sum(x**2), (2*x).astype(np.float32)
```

Comment:
I was considering whether this function is actually valid to pass, because it's a little odd at least to have different dtypes for the two return values. However, if the first return value is also `float32`, the failure mode remains the same - and then it's definitely a valid/normal function. So consider tweaking this test like so:
```python
def test_gh_issue18730():
    def fun(x):
        # A function that does things only in float32
        x = x.astype(np.float32)
        return np.sum(x**2), (2*x)
```